10x/count/count.wdl

```diff
@@ -8,20 +8,17 @@ task setup_chunks {
   String sample_id
   # See the adjacent sample_def.json.
   File sample_def
-  # Path to the directory that contains the fastq files
+  # Path to the tarball that contains the fastq files
   File read_path
   # auto seems to be the only reasonable value
   String chemistry_name = "auto"
 
   command <<<
     
-    # Move reads files to the cwd so strings can point to cwd rather than the full
-    # cromwell path
-    mkdir read_path
-    for read_file in '${read_path}'/*; do
-      ln "$read_file" read_path/"$(basename $read_file)"
-    done
-     
+    # note that reads must be tar'd in a flat directory (tar -cf files.tar *) for
+    # this unpacking to function properly.
+    mkdir read_path && tar -xf ${read_path} -C read_path/
```

Comment:
What about giving this task an array of fastqs instead, and then having this task move them to read_path? I suspect that real input bundles will have individual fastqs and not tarballs.

Reply:
On-point suggestion; This is something Marcus and I spoke about too. If we're getting these from cellranger mkfastq, I believe the input bundles should be directories with arbitrary numbers of fastqs (typically (1 to 8) * 3), so they're compatible with either approach. 

I elected to use the tarball because at the moment the pipeline processes these fastqs concurrently in a single wdl/cromwell call and because cellranger takes care of parsing the directory structure for us. 

As discussed this afternoon (I think just after this comment was made) there is an interest in breaking up earlier stages of the pipeline into single lanes and summarizing each lane with metrics. At that point it will *definitely* make sense to break this up into individual fastq files, as each wdl will take 3 fastqs (r1, r2, i1). 

I can definitely put in the change to fastq arrays; it will probably take 60-90 minutes to make the change + adjust the demo data and input files. Let me know if you think it's worth the time!

Reply:
I think the real type is `Array[Array[File]]` or like an `Array[Map[String, File]]`, though nesting of types seems to be disfavored by WDL.

But if this were to become an `Array[File]`, is there a preferred way to expand or iterate over that array in the command section? I keep using something like `'{sep="\' \'" var}'`, but maybe there's something better.

Reply:
I pushed a change here that uses this syntax:

```
  Array[File] r1
  Array[File] r2
  Array[File] i1
  # auto seems to be the only reasonable value
  String chemistry_name = "auto"

  command <<<

    # move all input files into read_path for cellranger
    mkdir read_path
    for fastq_file in ${sep=" " r1} ${sep=" " r2} ${sep=" " i1}; do
        mv $fastq_file read_path/
    done
```

Reply:
That should work for this. But in general if the file is named `my file` or `* my_favorite_fastq *` or `; do sleep 1; done; sleep infinity` (whatever), I think it breaks. I'm not certain about that though.

Also one other thing to consider: the interface of this task suggests that the r1, r2, and i1 files are going to be grouped as they're ordered in the input arrays. So you'd expect something like a `for triplet in zip(r1, r2, i1)` somewhere. But it really just puts everything in the same directory and uses filenames to infer relationships between the files.

Reply:
Regarding the first point: yes, the user could give us some malicious code and this is not secure (unless cromwell checks to see that it is actually a file). 

I agree completely with your second point. We do make those expectations, and would expect to see a zip somewhere. However, the main goal was to change the inputs so that the wrapper wdls and other green box machinery which rely on 10x inputs can pass "good" inputs. 

The way those inputs are *used* is still bad, but the files that interface with the pipelines will now be reusable when we change the internals of this pipeline to be more robust. 

Does that make sense? I'm trying to minimize time spent making this pipeline more robust since the consensus seemed to be "replace everything".

Reply:
Yeah, I'm mostly asking a general WDL question...so I can maybe do that in a venue other than this PR.

And I think it makes sense to delay some usability/robustness/etc discussions until we're somewhat closer to a workflow we're going to release.